torch_xla/csrc/convolution.cpp

```diff
@@ -316,6 +318,548 @@ ConvGrads BuildTransposedConvolutionBackward(
+  const int num_dims = num_spatial_dims + 2;
+  if (input_shape.rank() != num_dims) {
+    return tsl::errors::InvalidArgument(label, ": input must be ", num_dims,
+                                        "-dimensional");
+  }
+  if (filter_shape.rank() != num_dims) {
+    return tsl::errors::InvalidArgument(label, ": filter must be ", num_dims,
+                                        "-dimensional");
+  }
+  if (out_backprop_shape.rank() != num_dims) {
+    return tsl::errors::InvalidArgument(label, ": out_backprop must be ",
+                                        num_dims, "-dimensional");
+  }
+  int batch_dim = PTXLAGetTensorBatchDimIndex(num_dims, data_format);
+  dims->batch_size = input_shape.dimensions(batch_dim);
+  if (dims->batch_size != out_backprop_shape.dimensions(batch_dim)) {
+    return tsl::errors::InvalidArgument(
+        label, ": input and out_backprop must have the same batch size.",
+        " Input batch: ", dims->batch_size,
+        ", outbackprop batch: ", out_backprop_shape.dimensions(batch_dim),
+        ", batch_dim: ", batch_dim);
+  }
+
+  int feature_dim = PTXLAGetTensorFeatureDimIndex(num_dims, data_format);
+  dims->in_depth = input_shape.dimensions(feature_dim);
+  // The input and output feature dimensions are the second last and last
+  // dimensions of the filter Tensor.
+  VLOG(2) << "input vs filter_in depth " << dims->in_depth << " "
+          << filter_shape.dimensions(num_dims - 2);
+  if (filter_shape.dimensions(num_dims - 2) <= 0) {
+    return tsl::errors ::InvalidArgument(
+        label, ": filter depth must be strictly greated than zero");
+  }
+  if (dims->in_depth % filter_shape.dimensions(num_dims - 2)) {
+    return tsl::errors::InvalidArgument(
+        label, ": input depth must be evenly divisible by filter depth");
+  }
+  dims->out_depth = filter_shape.dimensions(num_dims - 1);
+  if (dims->out_depth != out_backprop_shape.dimensions(feature_dim)) {
+    return tsl::errors::InvalidArgument(
+        label, ": filter and out_backprop must have the same out_depth");
+  }
+  dims->spatial_dims.resize(num_spatial_dims);
+  for (int i = 0; i < num_spatial_dims; ++i) {
+    int image_dim = PTXLAGetTensorSpatialDimIndex(num_dims, data_format, i);
+    int64_t padding_before = -1, padding_after = -1;
+    if (padding == PTXLAPadding::EXPLICIT) {
+      padding_before = explicit_paddings[2 * image_dim];
+      padding_after = explicit_paddings[2 * image_dim + 1];
+    }
+    TF_RETURN_IF_ERROR(PTXLAConvBackpropExtractAndVerifyDimension(
+        label, input_shape, filter_shape, out_backprop_shape, dilations,
+        strides, padding, padding_before, padding_after, image_dim, i,
+        &dims->spatial_dims[i]));
+  }
+  return tsl::OkStatus();
+}
+
+// This code copied/inspired from TF:
+// https://github.com/tensorflow/tensorflow/blob/7f47eaf439d2b81de1aa24b10ed57eabd519dbdb/tensorflow/compiler/tf2xla/kernels/conv_op_helpers.cc#L52
+xla::PrecisionConfig PTXLAGetPrecisionConfig() {
```

Comment:
can you put all of the helper function in a separate `convolution_helper` file and group them by the original files? I think that way it is cleaner. Then you can say 
```
// -------------Convolution Helper Function Start-------------------------
// Convolution helper functions below are copied from TF2XLA bridge 
// This part of helpers are origionally from https://github.com/tensorflow/tensorflow/blob/7f47eaf439d2b81de1aa24b10ed57eabd519dbdb/tensorflow/core/util/tensor_format.cc#L40
```
```
// This part of helpers are origionally from 
// https://github.com/tensorflow/tensorflow/blob/7f47eaf439d2b81de1aa24b10ed57eabd519dbdb/tensorflow/compiler/tf2xla/kernels/conv_op_helpers.cc
```
etc

Reply:
Thanks, Jack, updated